File: workmanager_apple/ios/Classes/NotificationDebugHandler.swift
import Foundation
import UserNotifications

/**
 * A debug handler that shows notifications for task events.
 * Note: You need to ensure your app has notification permissions.
 */
public class NotificationDebugHandler: WorkmanagerDebug {
    private let identifier = UUID().uuidString
    private let workEmojis = ["👷‍♀️", "👷‍♂️"]
    private let successEmoji = "🎉"
    private let failureEmoji = "🔥"
    
    public override init() {}
    
    override func onTaskStatusUpdate(taskInfo: TaskDebugInfo, status: TaskStatus, result: TaskResult?) {
        let formatter = DateFormatter()
        formatter.dateStyle = .short
        formatter.timeStyle = .medium
        
        let (emoji, title, message) = formatNotification(taskInfo: taskInfo, status: status, result: result)
        
        scheduleNotification(
            title: "\(emoji) \(formatter.string(from: Date()))",
            body: "\(title)\n\(message)"
        )
    }
    
    override func onExceptionEncountered(taskInfo: TaskDebugInfo?, exception: Error) {
        let taskName = taskInfo?.taskName ?? "unknown"
        let formatter = DateFormatter()
        formatter.dateStyle = .short
        formatter.timeStyle = .medium
        
        scheduleNotification(
            title: "\(failureEmoji) \(formatter.string(from: Date()))",
            body: "Exception in Task\n• Task: \(taskName)\n• Error: \(exception.localizedDescription)"
        )
    }
    
    private func formatNotification(taskInfo: TaskDebugInfo, status: TaskStatus, result: TaskResult?) -> (String, String, String) {
        switch status {
        case .scheduled:
            return ("📅", "Task Scheduled", "• Task: \(taskInfo.taskName)\n• Input Data: \(taskInfo.inputData?.description ?? "none")")
        case .started:
            let workEmoji = workEmojis.randomElement() ?? "👷"
            return (workEmoji, "Task Starting", "• Task: \(taskInfo.taskName)\n• Callback Handle: \(taskInfo.callbackHandle ?? -1)")
        case .completed:
            let success = result?.success ?? false
            let duration = result?.duration ?? 0
            let emoji = success ? successEmoji : failureEmoji
            let title = success ? "Task Completed" : "Task Failed"
            var message = "• Task: \(taskInfo.taskName)\n• Duration: \(duration)ms"
            if let error = result?.error {
                message += "\n• Error: \(error)"
            }
            return (emoji, title, message)
        case .failed:
            let error = result?.error ?? "Unknown error"
            return (failureEmoji, "Task Failed", "• Task: \(taskInfo.taskName)\n• Error: \(error)")
        case .cancelled:
            return ("⚠️", "Task Cancelled", "• Task: \(taskInfo.taskName)")
        case .retrying:
            return ("🔄", "Task Retrying", "• Task: \(taskInfo.taskName)")
        }
    }
    
    private func scheduleNotification(title: String, body: String) {
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default
        
        let request = UNNotificationRequest(
            identifier: UUID().uuidString,
            content: content,
            trigger: nil // Immediate delivery
        )
        
        UNUserNotificationCenter.current().add(request) { error in
            if let error = error {
                print("Failed to schedule notification: \(error)")
            }
        }
    }
}

File: workmanager_android/android/src/main/kotlin/dev/fluttercommunity/workmanager/NotificationDebugHandler.kt
package dev.fluttercommunity.workmanager

import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.os.Build
import androidx.core.app.NotificationCompat
import dev.fluttercommunity.workmanager.pigeon.TaskStatus
import java.text.DateFormat
import java.util.Date
import java.util.concurrent.TimeUnit.MILLISECONDS
import kotlin.random.Random

/**
 * A debug handler that shows notifications for task events.
 * Note: You need to ensure your app has notification permissions.
 */
class NotificationDebugHandler : WorkmanagerDebug() {
    companion object {
        private const val DEBUG_CHANNEL_ID = "WorkmanagerDebugChannelId"
        private const val DEBUG_CHANNEL_NAME = "Workmanager Debug Notifications"
        private val debugDateFormatter =
            DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM)
    }

    private val workEmoji get() = listOf("👷‍♀️", "👷‍♂️").random()
    private val successEmoji = "🎉"
    private val failureEmoji = "🔥"
    private val warningEmoji = "⚠️"
    private val currentTime get() = debugDateFormatter.format(Date())

    override fun onTaskStatusUpdate(
        context: Context,
        taskInfo: TaskDebugInfo,
        status: TaskStatus,
        result: TaskResult?,
    ) {
        val notificationId = Random.nextInt()
        val (emoji, title, content) =
            when (status) {
                TaskStatus.SCHEDULED ->
                    Triple(
                        "📅",
                        "Task Scheduled",
                        "• Task: ${taskInfo.taskName}\n• Input Data: ${taskInfo.inputData ?: "none"}",
                    )
                TaskStatus.STARTED ->
                    Triple(
                        workEmoji,
                        "Task Starting",
                        "• Task: ${taskInfo.taskName}\n• Callback Handle: ${taskInfo.callbackHandle}",
                    )
                TaskStatus.COMPLETED -> {
                    val success = result?.success ?: false
                    val duration = MILLISECONDS.toSeconds(result?.duration ?: 0)
                    Triple(
                        if (success) successEmoji else failureEmoji,
                        if (success) "Task Completed" else "Task Failed",
                        "• Task: ${taskInfo.taskName}\n• Duration: ${duration}s${if (result?.error != null) "\n• Error: ${result.error}" else ""}",
                    )
                }
                TaskStatus.FAILED ->
                    Triple(
                        failureEmoji,
                        "Task Failed",
                        "• Task: ${taskInfo.taskName}\n• Error: ${result?.error ?: "Unknown error"}",
                    )
                TaskStatus.CANCELLED ->
                    Triple(
                        warningEmoji,
                        "Task Cancelled",
                        "• Task: ${taskInfo.taskName}",
                    )
                TaskStatus.RETRYING ->
                    Triple(
                        "🔄",
                        "Task Retrying",
                        "• Task: ${taskInfo.taskName}",
                    )
            }

        postNotification(
            context,
            notificationId,
            "$emoji $currentTime",
            "$title\n$content",
        )
    }

    override fun onExceptionEncountered(
        context: Context,
        taskInfo: TaskDebugInfo?,
        exception: Throwable,
    ) {
        val notificationId = Random.nextInt()
        val taskName = taskInfo?.taskName ?: "unknown"
        postNotification(
            context,
            notificationId,
            "$failureEmoji $currentTime",
            "Exception in Task\n• Task: $taskName\n• Error: ${exception.message}",
        )
    }

    private fun postNotification(
        context: Context,
        notificationId: Int,
        title: String,
        contentText: String,
    ) {
        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        createNotificationChannel(notificationManager)

        val notification =
            NotificationCompat
                .Builder(context, DEBUG_CHANNEL_ID)
                .setContentTitle(title)
                .setContentText(contentText)
                .setStyle(
                    NotificationCompat
                        .BigTextStyle()
                        .bigText(contentText),
                ).setSmallIcon(android.R.drawable.stat_notify_sync)
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .build()

        notificationManager.notify(notificationId, notification)
    }

    private fun createNotificationChannel(notificationManager: NotificationManager) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel =
                NotificationChannel(
                    DEBUG_CHANNEL_ID,
                    DEBUG_CHANNEL_NAME,
                    NotificationManager.IMPORTANCE_LOW,
                )
            notificationManager.createNotificationChannel(channel)
        }
    }
}
